File: simba-jdbc/src/test/java/me/ahoo/simba/jdbc/JdbcMutexContendServiceTest.java
/*
 * Copyright [2021-2021] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.ahoo.simba.jdbc;

import com.zaxxer.hikari.HikariDataSource;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import me.ahoo.simba.core.AbstractMutexContender;
import me.ahoo.simba.core.MutexOwner;
import me.ahoo.simba.core.MutexState;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

import static org.junit.jupiter.api.Assertions.*;

/**
 * @author ahoo wang
 */
@Slf4j
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class JdbcMutexContendServiceTest {
    private JdbcMutexOwnerRepository jdbcMutexOwnerRepository;
    private JdbcMutexContendServiceFactory contendServiceFactory;

    @BeforeAll
    void setup() {
        HikariDataSource hikariDataSource = new HikariDataSource();
        hikariDataSource.setJdbcUrl("jdbc:mysql://localhost:3306/simba_db");
        hikariDataSource.setUsername("root");
        hikariDataSource.setPassword("root");
        jdbcMutexOwnerRepository = new JdbcMutexOwnerRepository(hikariDataSource);
        contendServiceFactory = new JdbcMutexContendServiceFactory(jdbcMutexOwnerRepository, Duration.ofSeconds(2), Duration.ofSeconds(2), Duration.ofSeconds(5));
    }

    @Test
    void start() {
        String mutex = "start";
        CompletableFuture<MutexOwner> acquiredFuture = new CompletableFuture();
        CompletableFuture<MutexOwner> releasedFuture = new CompletableFuture();

        JdbcMutexContendService contendService = (JdbcMutexContendService) contendServiceFactory.createMutexContendService(new AbstractMutexContender(mutex) {
            @Override
            public void onAcquired(MutexState mutexState) {
                log.info("onAcquired");
                acquiredFuture.complete(mutexState.getAfter());
            }

            @Override
            public void onReleased(MutexState mutexState) {
                log.info("onReleased");
                releasedFuture.complete(mutexState.getAfter());
            }
        });

        contendService.start();
        acquiredFuture.join();
        assertTrue(contendService.isOwner());
        contendService.stop();
        releasedFuture.join();
        assertFalse(contendService.isOwner());

    }


    @Test
    void restart() {
        String mutex = "restart";
        jdbcMutexOwnerRepository.tryInitMutex(mutex);
        CompletableFuture<MutexOwner> acquiredFuture = new CompletableFuture();
        CompletableFuture<MutexOwner> releasedFuture = new CompletableFuture();

        CompletableFuture<MutexOwner> acquiredFuture2 = new CompletableFuture();
        CompletableFuture<MutexOwner> releasedFuture2 = new CompletableFuture();
        JdbcMutexContendService contendService = (JdbcMutexContendService) contendServiceFactory.createMutexContendService(new AbstractMutexContender(mutex) {
            @Override
            public void onAcquired(MutexState mutexState) {
                log.info("onAcquired");
                if (!acquiredFuture.isDone()) {
                    acquiredFuture.complete(mutexState.getAfter());
                } else {
                    acquiredFuture2.complete(mutexState.getAfter());
                }
            }

            @Override
            public void onReleased(MutexState mutexState) {
                log.info("onReleased");
                if (!releasedFuture.isDone()) {
                    releasedFuture.complete(mutexState.getAfter());
                } else {
                    releasedFuture2.complete(mutexState.getAfter());
                }
            }
        });

        contendService.start();
        acquiredFuture.join();
        assertTrue(contendService.isOwner());
        contendService.stop();
        releasedFuture.join();
        assertFalse(contendService.isOwner());

        contendService.start();
        acquiredFuture2.join();
        assertTrue(contendService.isOwner());
        contendService.stop();
        releasedFuture2.join();
        assertFalse(contendService.isOwner());
    }

    @SneakyThrows
    @Test
    void guard() {
        String mutex = "guard";
        jdbcMutexOwnerRepository.tryInitMutex(mutex);
        CompletableFuture<MutexOwner> acquiredFuture = new CompletableFuture();
        CompletableFuture<MutexOwner> releasedFuture = new CompletableFuture();

        JdbcMutexContendService contendService = (JdbcMutexContendService) contendServiceFactory.createMutexContendService(new AbstractMutexContender(mutex) {
            @Override
            public void onAcquired(MutexState mutexState) {
                log.info("onAcquired");
                acquiredFuture.complete(mutexState.getAfter());
            }

            @Override
            public void onReleased(MutexState mutexState) {
                log.info("onReleased");
                releasedFuture.complete(mutexState.getAfter());
            }
        });

        contendService.start();
        acquiredFuture.join();
        assertTrue(contendService.isOwner());
        TimeUnit.SECONDS.sleep(3);
        assertEquals(contendService.getAfterOwner().getOwnerId(), contendService.getContender().getContenderId());
        assertTrue(contendService.isOwner());
        contendService.stop();
        releasedFuture.join();
        assertFalse(contendService.isOwner());
    }

    @Test
    void multiContend() throws InterruptedException {
        String mutex = "multiContend";
        jdbcMutexOwnerRepository.tryInitMutex(mutex);
        AtomicInteger count = new AtomicInteger(0);
        AtomicReference<String> currentOwnerIdRef = new AtomicReference<>();
        List<JdbcMutexContendService> contendServiceList = new ArrayList<>(10);
        for (int i = 0; i < 10; i++) {
            JdbcMutexContendService contendService = (JdbcMutexContendService) contendServiceFactory.createMutexContendService(new AbstractMutexContender(mutex) {
                @Override
                public void onAcquired(MutexState mutexState) {
                    currentOwnerIdRef.set(mutexState.getAfter().getOwnerId());
                    super.onAcquired(mutexState);
                    assertEquals(1, count.incrementAndGet());
                }

                @Override
                public void onReleased(MutexState mutexState) {
                    super.onReleased(mutexState);
                    assertEquals(0, count.decrementAndGet());
                }
            });
            contendService.start();
            contendServiceList.add(contendService);
        }
        TimeUnit.SECONDS.sleep(30);
        assertEquals(1, count.get());
        String currentOwnerId = currentOwnerIdRef.get();
        for (JdbcMutexContendService contendService : contendServiceList) {
            if (contendService.getAfterOwner().getOwnerId() != null){
                Assertions.assertEquals(contendService.getAfterOwner().getOwnerId(), currentOwnerId);
            }
        }
        assertEquals(1, contendServiceList.stream().filter(contendService -> contendService.getContenderId().equals(currentOwnerId)).count());
    }
}

File: simba-jdbc/src/test/kotlin/me/ahoo/simba/jdbc/JdbcMutexContendServiceTest.kt
/*
 * Copyright [2021-present] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package me.ahoo.simba.jdbc

import com.zaxxer.hikari.HikariDataSource
import me.ahoo.simba.core.AbstractMutexContender
import me.ahoo.simba.core.MutexContendServiceFactory
import me.ahoo.simba.core.MutexState
import me.ahoo.simba.test.MutexContendServiceSpec
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import java.time.Duration
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

/**
 * @author ahoo wang
 */
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
internal class JdbcMutexContendServiceTest : MutexContendServiceSpec() {

    private lateinit var jdbcMutexOwnerRepository: JdbcMutexOwnerRepository
    override lateinit var mutexContendServiceFactory: MutexContendServiceFactory

    @BeforeAll
    fun setup() {
        val hikariDataSource = HikariDataSource()
        hikariDataSource.jdbcUrl = "jdbc:mysql://localhost:3306/simba_db"
        hikariDataSource.username = "root"
        hikariDataSource.password = "root"
        jdbcMutexOwnerRepository = JdbcMutexOwnerRepository(hikariDataSource)
        mutexContendServiceFactory = JdbcMutexContendServiceFactory(
            mutexOwnerRepository = jdbcMutexOwnerRepository,
            initialDelay = Duration.ofSeconds(2),
            ttl = Duration.ofSeconds(2),
            transition = Duration.ofSeconds(5)
        )

        jdbcMutexOwnerRepository.tryInitMutex(START_MUTEX)
        jdbcMutexOwnerRepository.tryInitMutex(RESTART_MUTEX)
        jdbcMutexOwnerRepository.tryInitMutex(GUARD_MUTEX)
        jdbcMutexOwnerRepository.tryInitMutex(MULTI_CONTEND_MUTEX)
        jdbcMutexOwnerRepository.tryInitMutex(SCHEDULE_MUTEX)
    }

    @Test
    fun `old owner is revoked before another contender acquires after renewal failure`() {
        val mutex = "renewal-failure-${System.nanoTime()}"
        jdbcMutexOwnerRepository.ensureOwner(mutex)
        val ownerAAcquired = CountDownLatch(1)
        val ownerAReleased = CountDownLatch(1)
        val ownerBAcquired = CountDownLatch(1)
        val attempts = AtomicInteger()
        val eventSequence = AtomicInteger()
        val ownerAReleasedAt = AtomicInteger()
        val ownerBAcquiredAt = AtomicInteger()
        val partitionedRepository = object : MutexOwnerRepository by jdbcMutexOwnerRepository {
            override fun acquireAndGetOwner(
                mutex: String,
                contenderId: String,
                ttl: Long,
                transition: Long
            ): MutexOwnerEntity {
                if (attempts.getAndIncrement() == 0) {
                    return jdbcMutexOwnerRepository.acquireAndGetOwner(mutex, contenderId, ttl, transition)
                }
                throw IllegalStateException("owner A cannot renew")
            }
        }
        val ownerA = object : AbstractMutexContender(mutex, "owner-a") {
            override fun onAcquired(mutexState: MutexState) {
                ownerAAcquired.countDown()
            }

            override fun onReleased(mutexState: MutexState) {
                ownerAReleasedAt.set(eventSequence.incrementAndGet())
                ownerAReleased.countDown()
            }
        }
        val ownerB = object : AbstractMutexContender(mutex, "owner-b") {
            override fun onAcquired(mutexState: MutexState) {
                ownerBAcquiredAt.set(eventSequence.incrementAndGet())
                ownerBAcquired.countDown()
            }
        }
        val ownerAService = JdbcMutexContendService(
            ownerA,
            Runnable::run,
            partitionedRepository,
            Duration.ZERO,
            Duration.ofMillis(200),
            Duration.ofMillis(100)
        )
        val ownerBService = JdbcMutexContendService(
            ownerB,
            Runnable::run,
            jdbcMutexOwnerRepository,
            Duration.ZERO,
            Duration.ofMillis(200),
            Duration.ofMillis(100)
        )

        try {
            ownerAService.start()
            assertThat("owner A should acquire first", ownerAAcquired.await(2, TimeUnit.SECONDS))
            ownerBService.start()
            assertThat("owner A should revoke local ownership", ownerAReleased.await(2, TimeUnit.SECONDS))
            assertThat("owner B should acquire after A's lease expires", ownerBAcquired.await(5, TimeUnit.SECONDS))

            assertThat(ownerAReleasedAt.get() < ownerBAcquiredAt.get(), equalTo(true))
            assertThat(ownerAService.isOwner, equalTo(false))
            assertThat(ownerBService.isOwner, equalTo(true))
        } finally {
            if (ownerBService.running) {
                ownerBService.stop()
            }
            if (ownerAService.running) {
                ownerAService.stop()
            }
            jdbcMutexOwnerRepository.release(mutex, "owner-a")
            jdbcMutexOwnerRepository.release(mutex, "owner-b")
        }
    }
}
